internal/querynode/shard_segment_detector.go

```diff
@@ -0,0 +1,209 @@
+// Licensed to the LF AI & Data foundation under one
+// or more contributor license agreements. See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership. The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License. You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package querynode
+
+import (
+	"context"
+	"sync"
+
+	"github.com/golang/protobuf/proto"
+	"github.com/milvus-io/milvus/internal/log"
+	"github.com/milvus-io/milvus/internal/proto/querypb"
+	"go.etcd.io/etcd/api/v3/mvccpb"
+	v3rpc "go.etcd.io/etcd/api/v3/v3rpc/rpctypes"
+	clientv3 "go.etcd.io/etcd/client/v3"
+	"go.uber.org/zap"
+)
+
+// etcdShardSegmentDetector watch etcd prefix for segment event.
+type etcdShardSegmentDetector struct {
```

Comment:
etcd shard detector may have lag?
When handoff happens, should the shard leader notice SegmentDetector before etcd notification?

Reply:
Etcd changes shall happen only after the segment is loaded. So the shard leader is guaranteed the segment is available after the segment is online. Then segment is release by `segmentsChangeInfo`. The only problem is the same data may be query/search twice, which will not affect the integrity.
For now, the `querycoord` still handle the hand-off management. This shall be improved in the next PRs that hand-off and load balance tasks will be sent to shard leaders and let them handles the segment states migration process.(which reference counts mechanism).